File: internal/capture/curl.go
package capture

import (
	"fmt"
	"sort"
	"strings"
)

// Shell 类型：决定转义与换行风格（2026-09-10 起只支持 cmd/bash，PowerShell 已删）
const (
	ShellCmd  = "cmd"  // Windows cmd：^" 包裹，对齐 Chrome DevTools escapeStringWin
	ShellBash = "bash" // bash / zsh / Git Bash：单引号/ANSI-C，对齐 DevTools escapeStringPosix
)

// MaxInlineBody cURL --data-raw 内联 body 上限（64KB，方案 §4.9）；超限或二进制不内联。
// 前端 frontend/src/lib/curl.ts 共用同一阈值与判定，改规则时两侧必须同步。
const MaxInlineBody = 64 << 10

// CurlResult BuildCurl 产物
type CurlResult struct {
	Command     string `json:"command"`
	BodyOmitted bool   `json:"bodyOmitted"` // body 二进制或超 64KB 未内联（提示从详情复制 Body）
}

// skipCurlHeaders 不输出到 cURL 的首部（小写比较）：代理逐跳/由 curl 自动维护头 +
// accept-encoding（对齐 Chrome DevTools ignoredHeaders）。
// Connection/Proxy-Connection/Keep-Alive/Transfer-Encoding/Upgrade 代理层记录前已剥离
// （见 proxy/server.go removeHopHeaders），此处保留仅为防御；host 由 curl 按 URL 补，
// content-length 让 curl 按 --data-raw 自动算，accept-encoding 让 curl 自己协商。
var skipCurlHeaders = map[string]bool{
	"connection":        true,
	"proxy-connection":  true,
	"keep-alive":        true,
	"transfer-encoding": true,
	"host":              true,
	"content-length":    true,
	"accept-encoding":   true,
}

// BuildCurl 按 Flow 请求生成可直接粘贴执行的 cURL 命令，形态对齐 Chrome 152 DevTools
// 「Copy as cURL」（前端 lib/curl.ts 同规则，改规则时两侧同步）：
//   - 首段裸 `curl --url <URL>`（可执行名不加引号；URL 内 {} [] 加反斜杠防 globbing）；
//   - 方法推断基于「是否真输出 data 段」：有内联 body→POST，否则 GET，不一致才输出 -X
//     （二进制/超限 body 被省略时，POST 会显式补 -X POST，否则 curl 误发 GET）；
//   - 头按 key ASCII 字典序；cookie 且值含 '=' 改用 -b（无 '=' 回退 -H 防 curl 读文件）；
//     值纯空白输出 -H 'k;'；
//   --data-raw 压尾；参数段 ≥3 才多行（cmd ' ^\n  ' / bash ' \\\n  '，两空格缩进），否则单行。
//
// body 为原始捕获字节（不解压）；无 NUL 且 ≤64KB 且未截断才内联，否则跳过并置 BodyOmitted。
func BuildCurl(f *Flow, shell string) (*CurlResult, error) {
	if f == nil || f.Request == nil {
		return nil, fmt.Errorf("该流无请求（盲透传隧道无明文请求），无法生成 cURL")
	}
	req := f.Request
	if req.URL == "" {
		return nil, fmt.Errorf("请求 URL 为空，无法生成 cURL")
	}
	if shell != ShellCmd && shell != ShellBash {
		return nil, fmt.Errorf("shell 须为 %q | %q", ShellCmd, ShellBash)
	}

	var quote func(string) string
	if shell == ShellCmd {
		quote = quoteCmd
	} else {
		quote = quoteBash
	}

	// 先判定 body 能否内联：方法推断依据「是否有 data 段」，必须在推断前算好。
	result := &CurlResult{}
	inlineBody := ""
	hasBody := false
	if len(req.Body) > 0 {
		hasBody = true
		if isText(req.Body) && len(req.Body) <= MaxInlineBody && !req.BodyTruncated {
			inlineBody = string(req.Body) // 原始 UTF-8 字节，非法序列按 RuneError 保留
		} else {
			result.BodyOmitted = true
		}
	}

	lines := []string{"curl --url " + escapeCurlURL(quote(req.URL))}

	method := strings.ToUpper(req.Method)
	inferred := "GET"
	if hasBody && !result.BodyOmitted {
		inferred = "POST"
	}
	if method != "" && method != inferred {
		lines = append(lines, "-X "+quote(method))
	}

	keys := make([]string, 0, len(req.Header))
	for k := range req.Header {
		if !skipCurlHeaders[strings.ToLower(k)] {
			keys = append(keys, k)
		}
	}
	sort.Strings(keys)
	for _, k := range keys {
		for _, v := range req.Header[k] {
			if strings.TrimSpace(v) == "" {
				// 空白值：curl 语法中分号表示空头
				lines = append(lines, "-H "+quote(k+";"))
				continue
			}
			if strings.EqualFold(k, "Cookie") && strings.Contains(v, "=") {
				lines = append(lines, "-b "+quote(v))
				continue
			}
			lines = append(lines, "-H "+quote(k+": "+v))
		}
	}

	if hasBody && !result.BodyOmitted {
		// 两平台均用 --data-raw：原样保留反斜杠等，不做 @ 文件展开
		lines = append(lines, "--data-raw "+quote(inlineBody))
	}

	sep := " "
	if len(lines) >= 3 {
		if shell == ShellCmd {
			sep = " ^\n  "
		} else {
			sep = " \\\n  "
		}
	}
	result.Command = strings.Join(lines, sep)
	return result, nil
}

// cmdWhitelist 为 cmd 转义中无需加 ^ 的可打印 ASCII（对齐 DevTools 白名单并含 ()：
// main 源码白名单不含括号，但 Chrome 152 实测 UA 中括号无 ^，以真实样例为准）。
const cmdWhitelist = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 _-:=+~/'.,?;*()"

// quoteCmd 逐字节移植 Chrome DevTools escapeStringWin（顺序不可调换）：
//  1. \ → \\、" → \"（先过 MS Crt 解析器）；
//  2. 白名单之外的每个 ASCII 可打印字符（\x21-\x7e）前加 ^（过 cmd.exe 解析器），
//     故 "→^\^"、{→^{、|→^|、`→^`；
//  3. 后跟字母/数字/下划线的 % 改 %^，避免 MS Crt 当环境变量展开；
//  4. 其余控制字符（含 TAB/DEL）→ 空格防命令注入；换行→ ^ 加两个换行（cmd 续行转义）；
//  5. 整体 ^"..."^ 包裹。
//
// 与 DevTools 唯一刻意差异：非 ASCII（>=0x80，如中文）原样保留，不加 ^ 也不替换空格
// ——替换会毁掉中文重放；现代 Windows Terminal/cmd 按活动代码页仍可粘贴执行。
func quoteCmd(s string) string {
	// 1) 反斜杠与双引号
	s = strings.ReplaceAll(s, `\`, `\\`)
	s = strings.ReplaceAll(s, `"`, `\"`)

	// 2) 白名单外的可打印 ASCII 前加 ^（逐字节，>=0x80 不动）
	var b strings.Builder
	b.Grow(len(s) + 8)
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c >= 0x21 && c <= 0x7e && !strings.ContainsRune(cmdWhitelist, rune(c)) {
			b.WriteByte('^')
		}
		b.WriteByte(c)
	}
	s = b.String()

	// 3) 后跟标识符字符的 % 防环境变量展开（用加 ^ 前的下一字节判定，^ 不是标识符）
	b.Reset()
	b.Grow(len(s) + 4)
	for i := 0; i < len(s); i++ {
		b.WriteByte(s[i])
		if s[i] == '%' && i+1 < len(s) && isIdentByte(s[i+1]) {
			b.WriteByte('^')
		}
	}
	s = b.String()

	// 4) 控制字符（含 TAB/DEL，不含换行）→ 空格
	b.Reset()
	b.Grow(len(s))
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c < 0x20 && c != '\n' && c != '\r' || c == 0x7f {
			b.WriteByte(' ')
		} else {
			b.WriteByte(c)
		}
	}
	s = b.String()

	// 换行→ ^ 加两个换行（\r\n 中的 \r 随后变 ^，整体与 DevTools 形态一致）
	s = strings.ReplaceAll(s, "\r\n", "^\n\n")
	s = strings.ReplaceAll(s, "\r", "^\n\n")
	s = strings.ReplaceAll(s, "\n", "^\n\n")
	return `^"` + s + `^"`
}

func isIdentByte(c byte) bool {
	return c == '_' || c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' || c >= '0' && c <= '9'
}

// quoteBash 移植 DevTools escapeStringPosix（按 rune 处理，避免 UTF-8 续接字节
// 0x80-0x9f 被误判为控制字符）：含控制字符/DEL-0x9f/!/单引号时用 ANSI-C 引用
// $'...'（!→\u0021 防历史展开、控制字符→\uXXXX，\n/\r 用 \n/\r），否则普通单引号。
func quoteBash(s string) string {
	needsAnsiC := strings.ContainsRune(s, '!') || strings.ContainsRune(s, '\'')
	if !needsAnsiC {
		for _, r := range s {
			if r < 0x20 || r >= 0x7f && r <= 0x9f {
				needsAnsiC = true
				break
			}
		}
	}
	if !needsAnsiC {
		return "'" + s + "'"
	}

	var b strings.Builder
	b.Grow(len(s) + 8)
	b.WriteString("$'")
	for _, r := range s {
		switch {
		case r == '\\':
			b.WriteString(`\\`)
		case r == '\'':
			b.WriteString(`\'`)
		case r == '\n':
			b.WriteString(`\n`)
		case r == '\r':
			b.WriteString(`\r`)
		case r == '!':
			b.WriteString(`\u0021`)
		case r < 0x20 || r >= 0x7f && r <= 0x9f:
			b.WriteString(fmt.Sprintf(`\u%04x`, r))
		default:
			b.WriteRune(r)
		}
	}
	b.WriteString("'")
	return b.String()
}

// escapeCurlURL DevTools 在 quote 之后对 URL 统一处理：{ } [ ] 前加反斜杠防 curl globbing
// （两平台同一规则；cmd 里 [ 不在白名单会先被加 ^，最终恰为 ^\[，与 DevTools 样例一致）。
func escapeCurlURL(quoted string) string {
	var b strings.Builder
	for i := 0; i < len(quoted); i++ {
		if quoted[i] == '{' || quoted[i] == '}' || quoted[i] == '[' || quoted[i] == ']' {
			b.WriteByte('\\')
		}
		b.WriteByte(quoted[i])
	}
	return b.String()
}

// isText 判断 body 是否可内联为文本：含 NUL 即二进制（对齐前端 lib/curl.ts 判定）。
func isText(b []byte) bool {
	for i := 0; i < len(b); i++ {
		if b[i] == 0 {
			return false
		}
	}
	return true
}

File: frontend/src/lib/curl.ts
// 前端版 cURL 生成：复盘归档流不在实时内存、Go BuildCurl 无法直接用，
// 故在前端用已加载的详情/请求体拼装。
//
// cmd 输出已对齐 Chrome 152 DevTools「Copy as cURL (cmd)」格式（2026-09-10 用户指定样例拍板，
// 规则取自 devtools-frontend NetworkLogView.ts _asCurlCommand/escapeStringWin）：
//   ① 裸 curl 开头、首参 --url URL，URL 内 {} [] 加反斜杠防 curl globbing；
//   ② 方法仅在与推断方法（有请求体→POST，无请求体→GET）不一致时才输出 -X；
//   ③ 头按 key 字典序（Go DTO 是 map，JSON 序列化即 ASCII 字典序，与样例一致），
//      cookie（且值含 '='）改用 -b，无 '=' 回退 -H 防 curl 读文件，空白值输出 'k;'；
//   ④ 请求体放 --data-raw 压在最后；参数段 ≥3 才多行，分隔符 ' ^\n  '，否则单行；
//   ⑤ cmd 转义：\→\\、"→\"，再给白名单 [A-Za-z0-9\s_-:=+~/'.,?;*()] 之外的字符加 ^，
//      后跟标识符的 % 改 %^，整体 ^"..."^ 包裹（所以 "→^\^"、{→^{、|→^|）。
//   与 DevTools 唯一刻意差异：非 ASCII（如中文）不替换成空格——原样保留以保证重放成功，
//   现代 Windows Terminal/cmd 按活动代码页仍可粘贴执行。
//
// bash 转义移植 DevTools escapeStringPosix（含控制字符/!/单引号时用 ANSI-C $'...'，
// 否则普通单引号），形态同为多行（\ 续行、两空格缩进）。
// Go 侧 internal/capture/curl.go 已同步为本套规则（2026-09-10，主窗 FlowList 右键 + CLI flows curl；
// 两侧均只支持 cmd/bash，默认 cmd）。
// body 内联规则（文本且 ≤64KB 且未截断才内联）两侧共用，改规则时必须同步本文件与 Go 侧。

export type CurlShell = 'cmd' | 'bash'

export interface CurlInput {
  url: string
  method?: string
  headers?: Record<string, string[]>
  body?: Uint8Array
  bodyTruncated?: boolean
}

export interface CurlResult {
  command: string
  bodyOmitted: boolean
}

// 文本启发式：含 NUL 字节视为二进制，不内联
function isText(data: Uint8Array): boolean {
  for (let i = 0; i < data.length; i++) {
    if (data[i] === 0) return false
  }
  return true
}

// 内联请求体上限：64KB（对齐 Go 侧 MAX_INLINE_BODY）
const MAX_INLINE_BODY = 64 * 1024

// 不输出到 cURL 的首部：代理逐跳/自动维护头 + accept-encoding（对齐 DevTools ignoredHeaders）
// Connection/Proxy-Connection/Keep-Alive/Transfer-Encoding 代理层记录前已剥离（见 server.go），
// 此处保留仅为防御；host 由 curl 按 URL 自动补；content-length 让 curl 按 body 自动补。
const SKIP_HEADERS = new Set([
  'connection',
  'proxy-connection',
  'keep-alive',
  'transfer-encoding',
  'host',
  'content-length',
  'accept-encoding',
])

// cmd 转义（逐字节移植 Chrome DevTools escapeStringWin，顺序不可调换）：
//   1) \→\\、"→\"（先过 MS Crt 解析器）；
//   2) 白名单 [A-Za-z0-9\s_-:=+~/'.,?;*()] 之外的每个字符前加 ^（过 cmd.exe 解析器），
//      故 "→^\^"、{→^{、|→^|、`→^`；白名单含 ()（Chrome 152 实测，用户样例 UA 中括号无 ^）；
//   3) 后跟字母/数字/下划线的 % 改成 %^，避免 MS Crt 当环境变量展开；
//   4) 其余控制字符（含 TAB）→空格，防命令注入；换行→^加两个换行（cmd 续行转义）；
//   5) 整体 ^"..."^ 包裹。
// 与 DevTools 唯一刻意差异：第 4 步不替换非 ASCII（中文等原样保留，保证重放成功）。
function quoteCmd(s: string): string {
  const r = s
    .replace(/\\/g, '\\\\')
    .replace(/"/g, '\\"')
    // 只给「白名单外的 ASCII（含 DEL）」加 ^；\u0080+ 非 ASCII（中文等）刻意原样保留（见文件头）
    .replace(/(?=[\x21-\x7e])[^a-zA-Z0-9\s_\-:=+~'\/.',?;*()]/g, '^$&')
    .replace(/%(?=[a-zA-Z0-9_])/g, '%^')
    .replace(/[\x00-\x09\x0b\x0c\x0e-\x1f\x7f]/g, ' ')
    .replace(/\r?\n|\r/g, '^\n\n')
  return '^"' + r + '^"'
}

// bash 转义（移植 DevTools escapeStringPosix）：含控制字符/DEL-0x9f/!/单引号时用
// ANSI-C 引用 $'...'（!→\u0021 防历史展开、控制字符→\uXXXX），否则普通单引号包裹。
function quoteBash(s: string): string {
  if (/[\0-\x1f\x7f-\x9f!]|'/.test(s)) {
    const toHex = (ch: string): string => '\\u' + ch.charCodeAt(0).toString(16).padStart(4, '0')
    return (
      "$'" +
      s
        .replace(/\\/g, '\\\\')
        .replace(/'/g, "\\'")
        .replace(/\n/g, '\\n')
        .replace(/\r/g, '\\r')
        .replace(/[\0-\x1f\x7f-\x9f!]/g, toHex) +
      "'"
    )
  }
  return "'" + s + "'"
}

// URL 额外转义（DevTools 在 escapeString 之后无分平台统一处理）：{} [] 前加 \ 防 curl globbing。
function escapeUrl(quoted: string): string {
  return quoted.replace(/[[{}\]]/g, '\\$&')
}

/**
 * 生成 cURL 命令（Chrome DevTools 形态）：首段 curl --url URL，其后 -X/-H/-b/--data-raw
 * 各占一段；参数段 ≥3 时多行（行尾 cmd `^` / bash `\` 续行，次行两空格缩进），否则单行。
 * 入参 url 为空时抛错，与 Go BuildCurl 的错误口径一致。
 */
export function buildCurl(input: CurlInput, shell: CurlShell): CurlResult {
  if (!input.url) {
    throw new Error('url is empty')
  }

  const quote = shell === 'cmd' ? quoteCmd : quoteBash

  // 按逻辑行收集：首行=可执行文件+--url，其后每段参数独占一行
  const lines: string[] = []

  // 可执行文件本身（curl）不含特殊字符，无需引号包裹
  lines.push('curl --url ' + escapeUrl(quote(input.url)))

  // 先判定 body 能否内联：DevTools 推断方法依据「是否有 data 段」——有内联 body→POST，否则 GET；
  // body 被省略（二进制/超限/截断）时必须显式 -X POST，否则 curl 会默认发 GET。
  let bodyOmitted = false
  let inlineBody: string | null = null
  if (input.body && input.body.length > 0) {
    if (isText(input.body) && input.body.length <= MAX_INLINE_BODY && !input.bodyTruncated) {
      inlineBody = new TextDecoder('utf-8', { fatal: false }).decode(input.body)
    } else {
      // 二进制/超限/已截断：不内联请求体，交由 UI 提示用户手工补
      bodyOmitted = true
    }
  }

  // 与推断方法不一致才输出 -X（对齐 DevTools）
  const method = (input.method || '').toUpperCase()
  const inferred = inlineBody !== null ? 'POST' : 'GET'
  if (method && method !== inferred) {
    lines.push('-X ' + quote(method))
  }

  // 首部按 key 字典序排序（对齐 DevTools，Go map JSON 序列化本身也是该顺序）
  const keys = Object.keys(input.headers || {}).filter((k) => !SKIP_HEADERS.has(k.toLowerCase())).sort()
  for (const k of keys) {
    for (const v of input.headers![k]) {
      // 值为空或纯空白的头输出 k;（curl 语法：分号表示空头）
      if (!v.trim()) {
        lines.push('-H ' + quote(k + ';'))
        continue
      }
      // cookie 仅在含 '=' 时用 -b；无 '=' 时 curl 会把值当 cookie 文件名读取（安全风险），回退 -H
      if (k.toLowerCase() === 'cookie' && v.includes('=')) {
        lines.push('-b ' + quote(v))
        continue
      }
      lines.push('-H ' + quote(k + ': ' + v))
    }
  }

  if (inlineBody !== null) {
    // DevTools 两平台均用 --data-raw（原样保留反斜杠等，不做 @ 文件展开）
    lines.push('--data-raw ' + quote(inlineBody))
  }

  // DevTools 规则：参数段 ≥3 才多行（分隔符 cmd ' ^\n  ' / bash ' \\\n  '，次行两空格缩进），
  // 否则单行空格连接；lines[0] 已含裸 curl 前缀。
  const sep = lines.length >= 3 ? (shell === 'cmd' ? ' ^\n  ' : ' \\\n  ') : ' '
  return { command: lines.join(sep), bodyOmitted }
}
